torch/functional.py

```diff
@@ -1439,6 +1440,42 @@ def norm(input, p="fro", dim=None, keepdim=False, out=None, dtype=None):  # noqa
         return handle_torch_function(
             norm, (input,), input, p=p, dim=dim, keepdim=keepdim, out=out, dtype=dtype)
 
+    # NB. All the repeated code and weird python is to please TorchScript.
+    #     For a more compact implementation see the relevant function in `_refs/__init__.py`
+
+    # We don't do this for MPS or sparse tensors
+    if input.layout == torch.strided and input.device.type in ("cpu", "cuda", "meta"):
+        if dim is not None:
+            if isinstance(dim, int):
+                _dim = [dim]
+            else:
+                _dim = dim
+        else:
+            _dim = None  # type: ignore[assignment]
+
+        if isinstance(p, str):
+            if p == "fro" and (dim is None or isinstance(dim, int) or len(dim) <= 2):
+                if out is None:
+                    return torch.linalg.vector_norm(input, 2, _dim, keepdim, dtype=dtype)
```

Comment:
I guess linalg.vector_norm and matrix_norm refs would need to be registered as decompositions to support `torch.norm` in functorch? cc @chillee, @zou3519 Are there functorch tests that would make sure that this doesn't break things?

Reply:
No i don't think this would be an issue with vmap. We already have a batching rule for `vector_norm`.

Reply:
According to [this failing test](https://github.com/pytorch/pytorch/runs/7469343187?check_suite_focus=true), we'll need to touch a bit some part of functorch, probably adding a batching rule for `matrix_norm`. Now that functorch is merged in core, I can do that in this PR directly.

Reply:
Yeah the functorch tests in PyTorch CI should give signal to this

Reply:
I added the decomposition for `matrix_norm`, but now I have a new error popping up:
https://github.com/pytorch/pytorch/runs/7862792767?check_suite_focus=true
I would xfail it, but I think it is relevant, as it may indicate that the decomposition is not correct? WDYT @zou3519?

Reply:
cc @samdow -- can we xfail this? matrix_norm requires the input to have two or more dimensions so I think this test doesn't apply

Reply:
So that test is checking that with a 1D input to the linalg functions, we get an error. So (1) it's good that we're still getting an error (🥳 ) and (2) it's weird that the error doesn't mention the `assertRaisesRegex` context manager that it's under

If you xfail it (and leave a comment that just says like "erroring but still fails test" please!), I'll look at it more since that's irrelevant to you

Reply:
It's not caught because the error type that's thrown `IndexError` does not match the error being caught. Sure, I'll xfail it for now. I think the fix should be quite simple for a person that understands functorch.

Reply:
FWIW, this xfail is not necessary anymore. Whatever was causing it has been fixed.

Reply:
Actually, it passes locally and in some CI jobs, and it fails in others. I'm skipping it.